devito/types/separable.py

```diff
@@ -0,0 +1,192 @@
+
+from devito.finite_differences import Differentiable
+from devito.finite_differences.differentiable import Mul, Add
+from devito.tools import as_tuple
+from devito.types.basic import AbstractCachedSymbol
+from devito.types.dense import Function, TimeFunction
+
+__all__ = ['SeparableFunction', 'SeparableTimeFunction']
+
+
+class SeparableFunction(Differentiable, AbstractCachedSymbol):
+    """
+    A Function separable in its Dimensions, ie f(x,y,z) = f(x)*f(y)*f(z)
+
+    Take the same parameters as a Function plus
+
+    Parameters
+    ----------
+    sep: str
+        Type of separation as a product or sum of Function, 'sum' or 'prod'.
+    separated: Dimension or tuple of Dimensions
+        Dimension in which the Function is separable.
+
+    Examples
+    --------
+    Creation
+
+    >>> from devito import Grid, TimeFunction
+    >>> grid = Grid(shape=(4, 4, 5))
+    >>> x, y, z = grid.dimensions
+    >>> SeparableFunction(name='a', grid=grid)
+    a_x(x)*a_y(y)*a_z(z)
+    >>> SeparableFunction(name='b', grid=grid, sep='sum')
+    b_x(x) + b_y(y) + b_z(z)
+    >>> SeparableFunction(name='c', grid=grid, sep='sum', separated=x)
+    c_x(x) + c_yz(y, z)
+    """
+    def __init__(self, *args, **kwargs):
+        # Is it sum or product separable
+        self._sep_op = spearation_op[kwargs.get('sep', 'prod')]
+
+        # Conventional Function inputs
+        self._grid = kwargs.get('grid')
+        self._name = kwargs.get('name')
+        self._space_order = kwargs.get('space_order', 1)
+
+        # Initialize subfunctions
+        # Separable dimensions
+        sep = self.separated_dimensions(**kwargs)
```

Comment:
for convention with the rest of the code, since this is a hook to be specialized by subclasses, I'd call it ``__setup_separated__``

Reply:
But isn't our convention that `__setup_separated__` would be a class method whil this is just an attribute

Reply:
that's how it started off, yes... then I had to extend it to instance methods too. See for example ``__halo_setup__`` and ``__padding_setup__`` -- these just can't be class methods. Same for ``__setup_staggered__`` I seem to remember.

Basically, the idea is that we use the convention ``__setup_X__`` to highlight (class, instance) methods that are expected to be specialized by subclasses.